### PR TITLE
Zerolog in Context

### PR DESCRIPTION
**Pull request by iamemilio**
Logging for the Go agent is here. This feature adds:
- Support for collecting log events and metrics
- The RecordLog() function for recording log data from a single log entry
- An integrated plugin for zero log to automatically ingest log data with the go agent

### CODE REVIEW

File: v3/newrelic/log_events.go [hunk 1]
```diff
@@ -133,11 +133,11 @@ func (events *logEvents) Merge(other *logEvents) {
 }
 
 func (events *logEvents) CollectorJSON(agentRunID string) ([]byte, error) {
-	if 0 == len(events.logs) {
+	if len(events.logs) == 0 {
 		return nil, nil
 	}
 
-	estimate := 500 * len(events.logs)
+	estimate := averageLogSizeEstimate * len(events.logs)
```
Comment (Ak-x) [inline v3/newrelic/log_events.go:140]: Appreciate the change here. 👍

File: v3/newrelic/log_event.go [hunk 2]
```diff
@@ -34,7 +35,12 @@ const (
 	MaxLogLength = 32768
 )
 
-// for internal user only
+// internal variable names and constants
+const (
+	// number of bytes expected to be needed for the average log message
+	averageLogSizeEstimate = 400
```
Comment (Ak-x) [inline v3/newrelic/log_event.go:41]: Good use of constants

File: v3/integrations/logcontext-v2/nrzerolog/Readme.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	baseLogger := zerolog.New(os.Stdout)
+
+	app, err := newrelic.NewApplication(
+		newrelic.ConfigFromEnvironment(),
+		newrelic.ConfigAppName("NRZerolog Example"),
+		newrelic.ConfigInfoLogger(os.Stdout),
+		newrelic.ConfigZerologPluginEnabled(true),
+		newrelic.ConfigDistributedTracerEnabled(true),
+	)
+	if nil != err {
```
Comment (iamemilio) [inline v3/integrations/logcontext-v2/nrzerolog/Readme.md:47]:
```suggestion
	if err != nil {
```

**Review by nr-swilloughby — DISMISSED** [review]

File: v3/examples/server/main.go [hunk 4]
```diff
@@ -251,6 +252,8 @@ func main() {
 		newrelic.ConfigAppName("Example App"),
 		newrelic.ConfigFromEnvironment(),
 		newrelic.ConfigDebugLogger(os.Stdout),
+		newrelic.ConfigDistributedTracerEnabled(true),
```
[the remark this answers is not in the document]
Comment (iamemilio) [reply v3/examples/server/main.go]: good point, we should remove that

File: v3/integrations/logcontext-v2/nrzerolog/Readme.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+## Installation
+
+The nrzerolog plugin, and the go-agent need to be integrated into your code
+in order to use this tool. Make sure to set `newrelic.ConfigZerologPluginEnabled(true)`
+in your config settings for the application. This will enable log forwarding and metrics
+in the go agent, as well as let the agent know that the zerolog pluging is in use.
+If you want to disable metrics, set `newrelic.ConfigAppLogMetricsEnabled(false),`.
+If you want to disable log forwarding, set `newrelic.ConfigAppLogForwardingEnabled(false),`.
+Note that the agent sets the default number of logs per harverst cycle to 10000, but that
+number may be reuced by the server. You can manually set this number by setting
```
Comment (RichVanderwal) [inline v3/integrations/logcontext-v2/nrzerolog/Readme.md:23]:
```suggestion
number may be reduced by the server. You can manually set this number by setting
```

File: v3/newrelic/log_event.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+	w.needsComma = false
+	buf.WriteByte(',')
+	w.intField(LogTimestampFieldName, e.timestamp)
+	buf.WriteByte('}')
+}
+
+// MarshalJSON is used for testing.
+func (e *logEvent) MarshalJSON() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, averageLogSizeEstimate))
```
Comment (RichVanderwal) [inline v3/newrelic/log_event.go:84]: Does hinting this with that "400" really make that much of a difference? Cool that we're thinking this hard about performance, though. I'd love to hear about your grouping of `int`s that I read earlier, too.

File: v3/integrations/logcontext-v2/nrzerolog/Readme.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+This plugin for zerolog implements the logs in context tooling for the go agent. This hook
+function can be added to any zerolog logger, and will automatically collect the log data
+from zerolog, and send it to New Relic through the go agent. The following Logging features
+are supported by this plugin in the current release:
+
+| Logging Feature | Supported |
+| ------- | --------- |
+| Forwarding | :heavy_check_mark: |
+| Metrics | :heavy_check_mark: |
+| Enrichment | :x: |
```
[the remark this answers is not in the document]
Comment (iamemilio) [reply v3/integrations/logcontext-v2/nrzerolog/Readme.md:12]: By enrichment, I mean that we do not support enriching the local logs that get printed to console in this release. We add all of the enrichments to the logs, including entity GUID, entity name, and hostname at harvest.

File: v3/newrelic/log_event.go (hunk 6 above)
Comment (iamemilio) [reply v3/newrelic/log_event.go:84]: Yes! The bytes buffer gets allocated on the heap, and when it grows, it has to estimate/request more memory. Each time it does that, we incur a cpu cost. By estimating a little more closely to what we think we need, we can cut down the number of buffer resizes we need, and save us cpu time, and run time!

Comment (nr-swilloughby) [reply v3/newrelic/log_event.go:84]: Yes, it's similar to C in that respect.

**Review by nr-swilloughby — DISMISSED** [review]
With the latest commits, this looks great. Lots of good, quality work here to keep performance and memory utilization optimized.

File: v3/examples/server/main.go [hunk 8]
```diff
@@ -274,6 +276,28 @@ func main() {
 	http.HandleFunc(newrelic.WrapHandleFunc(app, "/browser", browser))
 	http.HandleFunc(newrelic.WrapHandleFunc(app, "/async", async))
 	http.HandleFunc(newrelic.WrapHandleFunc(app, "/message", message))
+	http.HandleFunc("/log", func(w http.ResponseWriter, req *http.Request) {
+		app.RecordLog(newrelic.LogData{
```
Comment (RichVanderwal) [inline v3/examples/server/main.go:280]: Could we make this cleaner by writing a function to do the work? I like the tidiness of the HandleFuncs being here without the actual logic muddying it up.

File: v3/newrelic/log_events.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	buf.WriteByte(']')
+	return buf.Bytes(), nil
+}
+
+// split splits the events into two.  NOTE! The two event pools are not valid
+// priority queues, and should only be used to create JSON, not for adding any
+// events.
+func (events *logEvents) split() (*logEvents, *logEvents) {
+	// numSeen is conserved: e1.numSeen + e2.numSeen == events.numSeen.
+	sc1, sc2 := splitSeverityCount(events.severityCount)
```
Comment (RichVanderwal) [inline v3/newrelic/log_events.go:196]: Nice catch.

**Review by RichVanderwal — APPROVED** [review]
Great! This was a very large spec with lots of corners, and this code implements that with an eye to memory and cpu optimization to boot. The last-minute pivot in the interface was the right decision, and now the sample apps are simpler. Works in the UI!

**Pull request merged**